### ISSUE TITLE
Clippy doesn't report error/warnings after cargo check

### ISSUE THREAD

**Issue by b1zzu** [rust-lang/rust-clippy#4612]
## Environment

Project: https://github.com/b1zzu/wifiscanner/tree/clippy-bug (use the `clippy-bug` branch`)
Cargo: cargo 1.38.0 (23ef9a4ef 2019-08-20)
Clippy: clippy 0.0.212 (3aea860 2019-09-03)

## Steps to reproduce

Run `cargo clippy` after `cargo check`

```
cargo check
cargo clippy
```

Clippy doesn't fail or display any errors.

## Expected result

`cargo clippy` should fails with this errors:

```
warning: Constants have by default a `'static` lifetime
  --> src/lib.rs:87:22
   |
87 |     const PATH_ENV: &'static str = "PATH";
   |                     -^^^^^^^---- help: consider removing `'static`: `&str`
   |
   = note: `#[warn(clippy::redundant_static_lifetimes)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_static_lifetimes

error: approximate value of `f{32, 64}::consts::PI` found. Consider using it directly
  --> src/lib.rs:85:14
   |
85 |     let _x = 3.14;
   |              ^^^^
   |
   = note: `#[deny(clippy::approx_constant)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#approx_constant

warning: single-character string constant used as pattern
   --> src/lib.rs:157:16
    |
157 |         .split("\n")
    |                ^^^^ help: try using a char instead: `'\n'`
    |
    = note: `#[warn(clippy::single_char_pattern)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_char_pattern

warning: single-character string constant used as pattern
   --> src/lib.rs:168:36
    |
168 |     for line in network_list.split("\n") {
    |                                    ^^^^ help: try using a char instead: `'\n'`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_char_pattern

error: aborting due to previous error

error: Could not compile `wifiscanner`.

To learn more, run the command again with --verbose.
```
### Workaround

```
cargo check
cargo clean
cargo clippy
```

**Comment by mati865** [comment]
The only reasonable way to fix it is https://github.com/rust-lang/rust-clippy/issues/3837

With nightly toolchain you can use `cargo clippy-preview -Z unstable-options` as workaround.

**Comment by b1zzu** [comment]
@mati865 for now I will use cargo clean before clippy

**Comment by matthiaskrgr** [comment]
It works better if you `touch` all files of your root crate, this way you don't have to rebuild all the dependencies every time:

`find . | grep "\.rs$" | xargs touch ; cargo clippy`

**Comment by flip1995** [comment]
I think we should document this (`find . | grep "\.rs$" | xargs touch ; cargo clippy`) until `cargo clippy` (in `cargo`) is stable.

**Comment by b1zzu** [comment]
@flip1995 where would you document it? In a new separate Troubleshooting section or as a comment/tips to the [step-3-run-clippy](https://github.com/rust-lang/rust-clippy#step-3-run-clippy)

**Comment by flip1995** [comment]
I think at step-3-run-clippy is enough, since it will be removed anyway in the near future.

**Comment by maxbla** [comment]
This issue can also be triggered by running clippy multiple times with different options.
`main.rs`
```rust
fn main() {
    let _1 = 1_usize;
    println!("{}", _1);
}
```
```
$ cargo clean
$ cargo clippy -- --deny clippy::just_underscores_and_digits
    Compiling ...
error: consider choosing a more descriptive name
...
$ cargo clippy -- --allow clippy::all
$ cargo clippy -- --deny clippy::just_underscores_and_digits
```
`cargo clean` fixes this manifestation of the problem also

**Comment by sourcefrog** [comment]
It looks like this is now fixed? 
 
    cargo check
    cargo clippy

shows the right errors every time I run it.

    clippy 0.0.212 (69f99e7 2019-12-14)
    cargo 1.41.0 (626f0f40e 2019-12-03)

**Comment by flip1995** [comment]
That would surprise me, because we didn't change anything.

I just tested https://github.com/rust-lang/rust-clippy/issues/4612#issuecomment-563213742 and the behavior didn't change there.

**Comment by humb1t** [comment]
Clippy still depends on check: `clippy 0.0.212 (4ee1206 2020-02-01)`

**Comment by najamelan** [comment]
> The only reasonable way to fix it is #3837

What about having `cargo clippy` run the `touch **.rs` workaround itself? It would depend on platform support for the `touch` command and globbing, but I feel this issue is important enough to try this.

**Comment by mati865** [comment]
> What about having cargo clippy run the touch **.rs workaround itself?

You can just make a script that will call `touch` and Clippy.

**Comment by humb1t** [comment]
@mati865 don't you think that this is not a solution?
Imagine that `cargo test` will not run tests if `cargo build` was already executed.

**Comment by mati865** [comment]
@humb1t neither modifying files without permission is the solution.

**Comment by flip1995** [comment]
This is now fixed when running Clippy with 
```bash
cargo clippy -Zunstable-options
``` 
on nightly.

**Comment by humb1t** [comment]
@mati865 totally agree with you

**Comment by matthiaskrgr** [comment]
The feature should be available as soon as nightly 2020-04-18 rolls out.

**Comment by saleemrashid** [comment]
If I run the following commands (in this order) in my workspace, which has some `dead_code`, all the commands work as expected

```bash
# dead_code errors
cargo clippy -Z unstable-options -- -D warnings
# dead_code + clippy::restriction errors
cargo clippy -Z unstable-options -- -D warnings -D clippy::restriction
# dead_code errors
cargo clippy -Z unstable-options -- -D warnings
```

However, if I use warnings instead of errors, the first invocation's output is cached.

```bash
# dead_code warnings
cargo clippy -Z unstable-options
# dead_code warnings
cargo clippy -Z unstable-options -- -W clippy::restriction
find -name "*.rs" -exec touch {} \;
# dead_code + clippy::restriction warnings
cargo clippy -Z unstable-options -- -W clippy::restriction
# dead_code + clippy::restriction warnings
cargo clippy -Z unstable-options
```

---

```bash
rustc 1.44.0-nightly (b2e36e6c2 2020-04-22)
clippy 0.0.212 (891e1a8 2020-04-18)
```

**Comment by flip1995** [comment]
You should be using the nightly where this is already included. 

@yaahc any ideas?

**Comment by saleemrashid** [comment]
Here's a shorter reproduction case

```bash
echo "fn unused() {}" > src/lib.rs
cargo clippy -Z unstable-options
cargo clippy -Z unstable-options -- --allow dead_code
touch src/lib.rs
cargo clippy -Z unstable-options -- --allow dead_code
cargo clippy -Z unstable-options
```

![Screenshot](https://user-images.githubusercontent.com/22301423/80159618-09527980-85bb-11ea-8fee-710575c7c748.png)
![Screenshot](https://user-images.githubusercontent.com/22301423/80159642-11aab480-85bb-11ea-9b5f-3f02d8b95f58.png)

---

Note that it works fine with **errors**, but not warnings.

```bash
echo "pub fn bad() -> bool { 0 <= 0 }" > src/lib.rs
cargo clippy -Z unstable-options
cargo clippy -Z unstable-options -- --allow clippy::eq_op
touch src/lib.rs
cargo clippy -Z unstable-options -- --warn clippy::all
cargo clippy -Z unstable-options -- --allow clippy::eq_op
```

![Screenshot from 2020-04-24 00-37-16](https://user-images.githubusercontent.com/22301423/80159876-a2819000-85bb-11ea-96cb-6b5a18f643fe.png)
![Screenshot from 2020-04-24 00-37-48](https://user-images.githubusercontent.com/22301423/80159886-aa413480-85bb-11ea-83e4-6709f365737e.png)


```
rustc 1.44.0-nightly (b2e36e6c2 2020-04-22)
cargo 1.44.0-nightly (8751eb301 2020-04-21)
clippy 0.0.212 (891e1a8 2020-04-18)
```

**Comment by yaahc** [comment]
This is probably an issue with cargo, I'll check it out tomorrow.

cc @ehuss

**Comment by ehuss** [comment]
The caching with different flags is https://github.com/rust-lang/cargo/issues/7862.  The issue is twofold. One is that clippy hides the flags from cargo (via `__CLIPPY_HACKERY__`). The other is that Cargo doesn't provide a way to hash flags.

**Comment by retep998** [comment]
I'm running into this a lot because rust-analyzer will automatically run cargo check on save, so when I decide to run cargo clippy it doesn't report anything.

**Comment by flip1995** [comment]
@ehuss Is there a plan to stabilize the `RUSTC_WORKSPACE_WRAPPER ` env var in `cargo` in the near future, so that we can get this to stable? Since this is an issue, that comes up a few times a month, I'd be willing to put in the work, if there is still something to do before stabilization.

**Comment by ehuss** [comment]
There is an issue identified at https://github.com/rust-lang/cargo/issues/8143#issuecomment-619289546. I think it would be good to at least have a design plan (and preferably an implementation) to fix that before stabilizing, in case it influences how the wrapper should work.

**Comment by jonasmalacofilho** [comment]
Knowing that the problem is the reuse of metadata from regular/other `check` runs, I found that running `clippy` with a different `--target-dir` is sufficient to work around the problem.

```
$ cargo check
$ cargo clippy -- [clippy options]
[... nothing ...]
$ cargo clippy --target-dir target/clippy -- [clippy options]
[... clippy warnings and errors ...]
```

Note that I don't think this will necessarily do the right thing when workspaces are involved, particularly if `cargo clippy` is not invoked from the workspace root.

**Comment by lf-** [comment]
By the way, if anyone is having annoying issues with rust-analyzer running `cargo check` on save so your clippy doesn't work, you can do `cargo clippy --release` without having to introduce a separate feature.

**Comment by flip1995** [comment]
> Also, is this related to #2604?

Yes this is basically the same

**Comment by mikechambers** [comment]
Just want to report that I am still seeing this, even with the workarounds mentioned above (using nightly, and find). Seeing it on Linux and Windows with the December 21 nightly. Basically, clippy just no longer works.

**Comment by wpbrown** [comment]
As far as workarounds go, the [find](https://github.com/rust-lang/rust-clippy/issues/4612#issuecomment-537633718) workaround has been working for me. I put `find . -name '*.rs' | xargs touch` in [my precommit](https://github.com/blockcaptain/blockcaptain/blob/master/.pre-commit-config.yaml) to always run before clippy.

**Comment by mikechambers** [comment]
@wpbrown Thanks. Yes, I tried that (and setting the unstable flags). Neither was working (which is why i posted here in case it was a larger issue).

**Comment by vallentin** [comment]
People already supplied workarounds using `find` and `touch`. Let's just include one more variation for improving the overall performance, which is useful for big projects and workspaces.

- Instead of `grep "\.rs$"` then use `-name "*.rs"` as part of `find`
- Instead of `xargs touch` then use `-exec touch "{}" +` also as part of `find`
- Maybe even skip the `target/` dir by adding `-not -path "./target/*"` as part of `find`

```shell
find -name "*.rs" -not -path "./target/*" -exec touch "{}" +
```

Then lastly `cargo clippy` of course.

**Comment by Fishrock123** [comment]
This is a recurring pain point for the Rust team I work on.

To get clippy to lint, I usually have to `cargo clean` first.

**Comment by najamelan** [comment]
@Fishrock123 for me `touch **.rs` works and will be faster than clean. But yeah, it's an annoying bug since if fails silently pretending lints pass.

**Comment by camsteffen** [comment]
> (Why the heck was this downvoted? As I said, it's a recurring pain point for my team!!)

@Fishrock123 Because you're making the problem even worse by needlessly rebuilding the entire project.

**Comment by flip1995** [comment]
> (Why the heck was this downvoted? As I said, it's a recurring pain point for my team!!)

@Fishrock123 Posting the same again and again on an issue where already many workarounds were posted, gets downvoted on GitHub most of the time (and in open source in general). We know about this issue and are trying to get it fixed, it's just not as easy as it looks. It is even pinned in this repo. It doesn't help if every other week a comment saying "this is still an issue" without suggesting a new solution is posted on the issue.

I, personally, just ignore those comments (usually).

**Comment by flip1995** [comment]
This is now fixed in Clippy and will land in some nightly over the weekend. This means the fix will hit stable (if nothing goes wrong) in 1.52.0.

**Comment by tesuji** [comment]
Hi, I'm happy this issue fixed.
From [the changelog](https://github.com/rust-lang/rust-clippy/pull/6687#issue-568832165) it seems that 

> cargo clippy no longer shares the same build cache as cargo check.

One issue for me is that clippy run slower than cargo-check command.
Before this change, I run cargo check and touch main.rs or lib.rs to redo the lints by clippy.

There are some reasons why clippy is slower than cargo-check:
1) clippy has many more lints to run compare to just official lints supported by rust in cargo-check.
2) even though warnings/errors by lints for transitive dependencies doesn't show by default,
  the lints actually run for all the code in dependencies. 

Should I opened an issue to track this on clippy side ?

Those said, my personal issue should not be a reason to revert the fix.
Thank you for working on this.

**Comment by flip1995** [comment]
> One issue for me is that clippy run slower than cargo-check command.

Yes this is true for your crate, not for the dependencies. This is because Clippy runs more lints and analyzes your code more than `cargo check`. It shouldn't be that noticeable though.

Clippy lints are not run for dependencies (except local path dependencies, but you can disable this behavior with `cargo clippy -- --no-deps`). The issue you linked isn't really about linting dependencies - that's just an effect of this issue. There's nothing from the Clippy side that could be done.

> Before this change, I run cargo check and touch main.rs or lib.rs to redo the lints by clippy.

This change didn't even land yet. The effect of this change is solely that you don't have to `touch` those files anymore to run `cargo clippy` after `cargo check`.

**Comment by hellow554** [comment]
This issue is still pinned. Is this on purpose or just a leftover? I could understand it if you want people who aren't using the latest stable version to see this in the first place. But as said: is it on purpose?

**Comment by flip1995** [comment]
Since last Thursday it isn't on purpose anymore, since this hit stable :tada: Thanks again `@everyone` who worked on this! I removed the pin, thanks for reminder.
